test/client.js

```diff
@@ -316,5 +314,69 @@ describe('MqttClient', function () {
         })
       })
     })
+
+    it('should not send the same publish multiple times on a flaky connection', function (done) {
+      this.timeout(3500)
+
+      var KILL_COUNT = 4
+      var killedConnections = 0
+      var pubIds = {}
+      var client = mqtt.connect({
+        port: port + 47,
+        host: 'localhost',
+        connectTimeout: 350,
+        reconnectPeriod: 300
+      })
+
+      var server2 = net.createServer(function (stream) {
+        var client = new Connection(stream)
+        client.on('error', function () {})
+        client.on('connect', function (packet) {
+          if (packet.clientId === 'invalid') {
+            client.connack({returnCode: 2})
+          } else {
+            client.connack({returnCode: 0})
+          }
+        })
+
+        this.emit('client', client)
+      }).listen(port + 47)
+
+      server2.on('client', function (c) {
+        client.publish('topic', 'data', { qos: 1 }, function () {
+          done()
+          client.end(true)
+          c.destroy()
+          server2.destroy()
+        })
+
+        c.on('publish', function onPublish (packet) {
+          if (killedConnections < KILL_COUNT) {
+            // Kill the first few pub attempts to simulate a flaky connection
+            killedConnections++
+            c.destroy()
+
+            // to avoid receiving inflight messages
+            c.removeListener('publish', onPublish)
```

Comment:
because `c.destroy` destroys the stream/socket, but the packet has been already received and parsed, so it will be delivered synchronously.

Please verify on AWS IoT if everything works correctly with this fix.